server/docs/design/bidi-producer-consumers-streaming.md

```diff
@@ -0,0 +1,339 @@
+the `ResponseStreamObserver` managed by Helidon for transmitting BlockItemResponses to the downstream consumer.  The 
+`BlockItemStreamService` will also subscribe the `ConsumerBlockItemObserver` to the `StreamMediator` to receive the
+streaming BlockItems from the producer.
+
+
+### Runtime Streaming
+
+At runtime, the `ProducerBlockItemObserver` will receive the latest BlockItem from the producer via Helidon and will 
+invoke publishEvent(BlockItem) on the `StreamMediator` to write the BlockItem to the RingBuffer.  The 
+`ProducerBlockItemObserver` will then persist the BlockItem and return a BlockItemResponse to the producer via 
+its reference to `ResponseStreamObserver`.
+
+Asynchronously, the RingBuffer will invoke the onEvent(BlockItem) method of all the subscribed 
+`ConsumerBlockItemObserver`s passing them the latest BlockItem.  The `ConsumerBlockItemObserver` will then transmit
+the BlockItem downstream to the consumer via its reference to the `ResponseStreamObserver`.  Downstream consumers will
+respond with a BlockItemResponse.  Helidon will call the onNext() method of the `ConsumerBlockItemObserver` with the
+BlockItemResponse.
+
+BlockItems sent to the `ConsumerBlockItemObserver` via the RingBuffer and BlockItemResponses passed by Helidon from
+the downstream consumer are used to refresh internal timeouts maintained by the `ConsumerBlockItemObserver`.  If a
+configurable timeout threshold is exceeded, the `ConsumerBlockItemObserver` will unsubscribe itself from the 
+`StreamMediator`.  This mechanism is necessary because producers and consumers may not send HTTP/2 `End Stream` DATA 
+frames to terminate their bidirectional connection.  Moreover, Helidon does not throw an exception back up to 
+`ConsumerBlockItemObserver` when the downstream consumer disconnects.  Internal timeouts ensure objects are not
+permanently subscribed to the `StreamMediator`.
+
+### Entities
+
+**BlockItemStreamService** - The BlockItemStreamService is a custom implementation of the Helidon gRPC GrpcService.
+It is responsible for initializing the StreamMediator and instantiating ProducerBlockItemObserver and 
+ConsumerBlockItemObserver instances on-demand when the gRPC API is called by producers and consumers.  It is
+the primary binding between the Helidon routing mechanisms and the `hedera-block-node` custom business logic.
+
+**StreamObserver** - StreamObserver is the main interface through which Helidon 4.x.x invokes custom business logic
+to receive and transmit bidirectional BlockItem streams at runtime.
+
+**ProducerBlockItemObserver** - A custom implementation of StreamObserver invoked by Helidon at runtime which is
+responsible for:
+1) Receiving the latest BlockItem from the producer (e.g. Consensus Node).
+2) Returning a response to the producer.
+
+**StreamMediator** - StreamMediator is an implementation of the [Mediator Pattern](https://en.wikipedia.org/wiki/Mediator_pattern)
```

Comment:
I don't understand the value a mediator brings other than adding complexity?

Reply:
The `StreamMediator` seemed like a reasonable component to encapsulate details around the [Disruptor integration with the StreamObserver](https://github.com/hashgraph/hedera-block-node/blob/fef178deb4f4a22fd4135f370fe87cc74148f108/server/src/main/java/com/hedera/block/server/mediator/LiveStreamMediatorImpl.java#L77-L105). Without this abstraction, the `RingBuffer` would need to be anchored on the `BlockStreamService` and all subscribe/unsubscribe operations to the `RingBuffer` would be mixed with `GrpcService API` concerns. Do you object to the name? Maybe we could change that to something like `BroadcastStreamController`?

Reply:
I'll keep the `StreamMediator` for the time being unless there's additional feedback.